modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSource.java

```diff
@@ -0,0 +1,418 @@
+                    if (dvs.advanceExact(doc)) {
+                        long ord;
+                        while ((ord = dvs.nextOrd()) != NO_MORE_ORDS) {
+                            values[0] = ord;
+                            next.collect(doc, 0L);
+                        }
+                    }
+                }
+            };
+        }
+
+        private static long lookupGlobalOrdinals(GlobalOrdinalMapping mapping, BytesRef key) throws IOException {
+            long low = 0;
+            long high = mapping.getValueCount();
+
+            while (low <= high) {
+                long mid = (low + high) >>> 1;
+                BytesRef midVal = mapping.lookupOrd(mid);
+                int cmp = midVal.compareTo(key);
+                if (cmp < 0) {
+                    low = mid + 1;
+                } else if (cmp > 0) {
+                    high = mid - 1;
+                } else {
+                    return mid;
+                }
+            }
+            return low-1;
+        }
+    }
+
+    /**
+     * A {@link CompositeValuesSource} for binary source ({@link BytesRef})
+     */
+    private static class BinaryValuesSource extends CompositeValuesSource<ValuesSource.Bytes, BytesRef> {
+        private final BytesRef[] values;
+        private BytesRef topValue;
+
+        BinaryValuesSource(ValuesSource.Bytes vs, int size, int reverseMul) {
+            super(vs, size, reverseMul);
+            this.values = new BytesRef[size];
+        }
+
+        @Override
+        String type() {
+            return "binary";
+        }
+
+        @Override
+        public void move(int from, int to) {
+            values[to] = BytesRef.deepCopyOf(values[from]);
+        }
+
+        @Override
+        public int compare(int from, int to) {
+            return values[from].compareTo(values[to]) * reverseMul;
+        }
+
+        @Override
+        int compareTop(int slot) {
+            return values[slot].compareTo(topValue) * reverseMul;
+        }
+
+        @Override
+        void setTop(Comparable<?> value) {
+            if (value.getClass() == BytesRef.class) {
+                topValue = (BytesRef) value;
+            } else if (value.getClass() == String.class) {
+                topValue = new BytesRef((String) value);
+            } else {
+                throw new IllegalArgumentException("invalid value, expected string, got " + value.getClass().getSimpleName());
+            }
+        }
+
+        @Override
+        Comparable<BytesRef> toComparable(int slot) {
+            return values[slot];
+        }
+
+        @Override
+        LeafBucketCollector getLeafCollector(LeafReaderContext context, LeafBucketCollector next) throws IOException {
+            final SortedBinaryDocValues dvs = vs.bytesValues(context);
+            return new LeafBucketCollector() {
+                @Override
+                public void collect(int doc, long bucket) throws IOException {
+                    if (dvs.advanceExact(doc)) {
```

Comment:
The bucket variable is always zero here, right? If that is the case then maybe we should have assertions for this in these LeafBucketCollector anonymous classes?